frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,372 @@
+        Vue.set(state.messages, [state.messages.length], newMessage)
+
+        if (state.users[username] && !state.users[username].departedDate) {
+          console.log(`chatroom Join: ${username} is already joined the chatroom #${state.name}`)
+          return
+        }
+
+        Vue.set(state.users, username, {
+          joinedDate: meta.createdDate,
+          departedDate: null
+        })
+      },
+      sideEffect ({ contractID, hash }, { state }) {
+        emitMessageEvents({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
+      }
+    },
+    'gi.contracts/chatroom/rename': {
+      validate: objectOf({
+        name: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        Vue.set(state.attributes, 'name', data.name)
+        const notificationData = createNotificationData(MESSAGE_NOTIFICATIONS.UPDATE_NAME, {})
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        Vue.set(state.messages, [state.messages.length], newMessage)
+      },
+      sideEffect ({ contractID, hash }, { state }) {
+        emitMessageEvents({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
+      }
+    },
+    'gi.contracts/chatroom/changeDescription': {
+      validate: objectOf({
+        description: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        Vue.set(state.attributes, 'description', data.description)
+        const notificationData = createNotificationData(
+          MESSAGE_NOTIFICATIONS.UPDATE_DESCRIPTION, {}
+        )
+        const newMessage = createMessage({ meta, hash, data: notificationData, state })
+        Vue.set(state.messages, [state.messages.length], newMessage)
+      },
+      sideEffect ({ contractID, hash }, { state }) {
+        emitMessageEvents({ type: MESSAGE_ACTION_TYPES.ADD_MESSAGE, contractID, hash, state })
+      }
+    },
+    'gi.contracts/chatroom/leave': {
+      validate: objectOf({
+        username: string
+      }),
+      process ({ data, meta, hash }, { state }) {
+        const { username } = data
+        if (state.users[username] && !state.users[username].departedDate) {
+          if (!sbp('okTurtles.data/get', 'JOINING_CHATROOM')) {
+            Vue.set(state.users[username], 'departedDate', meta.createdDate)
+          }
+
+          const notificationType = username === meta.username ? MESSAGE_NOTIFICATIONS.LEAVE_MEMBER : MESSAGE_NOTIFICATIONS.KICK_MEMBER
+          const notificationData = createNotificationData(
+            notificationType,
+            notificationType === MESSAGE_NOTIFICATIONS.KICK_MEMBER ? { username } : {})
+          const newMessage = createMessage({ meta, hash, data: notificationData, state })
+          Vue.set(state.messages, [state.messages.length], newMessage)
+          return
+        }
+        console.log(`chatroom Leave: ${username} is not a member of this chatroom #${state.name}`)
```

Comment:
Is this something that should be impossible via the UI? If so, it's an error and should be handled as such. See comment above.

Reply:
Already implemented the possibility to leave in UI. So correct, we should handle this as an Exception here.